RecoParticleFlow/PFClusterProducer/interface/PFEcalEndcapRecHitCreator.h

```diff
@@ -0,0 +1,229 @@
+      iSetup.get<CaloGeometryRecord>().get(geoHandle);
+  
+      iEvent.getByToken(srFlagToken_,srFlagHandle_);
+
+      // get the ecal geometry
+      const CaloSubdetectorGeometry *gTmp = 
+	geoHandle->getSubdetectorGeometry(DetId::Ecal, EcalEndcap);
+
+      const EcalEndcapGeometry *ecalGeo =dynamic_cast< const EcalEndcapGeometry* > (gTmp);
+
+      iEvent.getByToken(recHitToken_,recHitHandle);
+      for(const auto& erh : *recHitHandle ) {      
+	const DetId& detid = erh.detid();
+	auto energy = erh.energy();
+	auto time = erh.time();
+        bool hi = isHighInterest(detid);
+
+	const CaloCellGeometry * thisCell= ecalGeo->getGeometry(detid);
+  
+	// find rechit geometry
+	if(!thisCell) {
+          throw cms::Exception("ECALCellNotInGeometry")
+	    << "Detid: "<< detid.rawId()<<" not found in geometry";
+	  continue;
+	}
+
+	out->emplace_back(thisCell, detid.rawId(),PFLayer::ECAL_ENDCAP,
+			   energy); 
+
+        auto & rh = out->back();
+	
+	bool rcleaned = false;
+	bool keep=true;
+
+	//Apply Q tests
+	for( const auto& qtest : qualityTests_ ) {
+	  if (!qtest->test(rh,erh,rcleaned,hi)) {
+	    keep = false;	    
+	  }
+	}
+	  
+	if(keep) {
+	  rh.setTime(time);
+	  rh.setDepth(1);
+	} 
+        else {
+	  if (rcleaned) 
+	    cleaned->push_back(std::move(out->back()));
+          out->pop_back();
+        }
+      }
+    }
+
+    void init(const edm::EventSetup &es) {
+
+      edm::ESHandle<CaloGeometry> pG;
+      es.get<CaloGeometryRecord>().get(pG);   
+
+      edm::ESHandle<EcalTrigTowerConstituentsMap> hetm;
+      es.get<IdealGeometryRecord>().get(hetm);
+      eTTmap_ = &(*hetm);
+  
+      const EcalEndcapGeometry * myEcalEndcapGeometry = dynamic_cast<const EcalEndcapGeometry*>(pG->getSubdetectorGeometry(DetId::Ecal,EcalEndcap));
+      const std::vector<DetId>& vec(myEcalEndcapGeometry->getValidDetIds(DetId::Ecal,EcalEndcap));
+      unsigned size=vec.size();
+      for(unsigned ic=0; ic<size; ++ic) 
+        {
+          EEDetId myDetId(vec[ic]);
+          int cellhashedindex=myDetId.hashedIndex();
+          // a bit of trigger tower and SuperCrystals algebra
+          // first get the trigger tower 
+          EcalTrigTowerDetId towid1= eTTmap_->towerOf(vec[ic]);
+          int tthashedindex=TThashedIndexforEE(towid1);
+          towerOf_[cellhashedindex]=tthashedindex;
+
+          // get the SC of the cell
+          int schi=SChashedIndex(EEDetId(vec[ic]));
+          if(schi<0) EEDetId myID(vec[ic]);
+      
+          theTTDetIds_[tthashedindex]=towid1;
+
+          // check if this SC is already in the list of the corresponding TT
+          std::vector<int>::const_iterator itcheck=find(SCofTT_[tthashedindex].begin(),
+                                                        SCofTT_[tthashedindex].end(),
+                                                        schi);
+          if(itcheck==SCofTT_[tthashedindex].end())
+            SCofTT_[tthashedindex].push_back(schi);
```

Comment:
Is the size of ```SCofTT_[tthashedindex]``` known? If it's known roughly, please add a reserve; if it's known perfectly and it's constant over tthashedindex,  better change the logic to use std::array